### PR TITLE
Issue #436: Automate release of freshly-built snap to latest/stable channel

### LINKED ISSUE #436: Automate release of freshly-built snap from latest/edge to latest/stable channel
A build in snapcraft.io runs automatically on each push to `master`.

After it's complete, the following shows up at https://snapcraft.io/git-machete/releases :

image

I still need to do the release from edge to stable manually from web or via snapcraft CLI.

I think this could be automated as well &mdash; either by:
1. turning off snapcraft.io builds & building the snap within CI & pushing to Snap Store from CI (probably harder to configure, easier to manage long-term), OR
2. by actively waiting for the builds in Snap Store to complete (easier to configure, harder to manage long-term).

### PR DESCRIPTION
**Pull request by amalota**
closes #436

### CODE REVIEW

File: ci/snap/ci-deploy.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# `--use-lxd` applied to use a LXD container instead of a VM, to work around lack of support for KVM on CircleCI VMs.
+snapcraft --use-lxd
+
+if [[ ${1-} == "--dry-run" || ${CIRCLE_BRANCH-} != "master" ]]; then
+  ! command -v git-machete
+  sudo snap install git-machete*.snap --dangerous --classic
+  git machete version
+  sudo snap remove git-machete
+  echo "MASTER"
+  echo "$SNAPCRAFT_LOGIN_CREDENTIALS_CONTENTS_BASE64" | base64 -d > ~/.snapcraft.login
```
Comment (PawelLipski) [inline ci/snap/ci-deploy.sh]: A security notice/warning here: while `$SNAPCRAFT_LOGIN_CREDENTIALS_CONTENTS_BASE64` will be masked in the CI logs (which are **public**), the un-base64-ed contents won't... AFAICS now they don't leak, but be wary of that (turning off `set -x` might be a good idea in the long run)

File: ci/snap/snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+name: git-machete2
```
Comment (PawelLipski) [inline ci/snap/snapcraft.yaml]: Is this file picked up now by the test CI script? 🤔 or rather snap/snapcraft.yaml

Comment (amalota) [reply ci/snap/snapcraft.yaml]: @PawelLipski Yes, it is picked up

Comment (amalota) [reply ci/snap/snapcraft.yaml]: Hmm, turns out it wasn't picked up, i didn't even know that there already is snapcraft.yaml file in repo :O

File: ci/snap/ci-deploy.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+#if [[ ${1-} == "--dry-run" || ${CIRCLE_BRANCH-} != "master" ]]; then
+if [[ ${1-} == "--dry-run" ]]; then
+  echo "DEVELOP"
+  ! command -v git-machete
+  sudo snap install git-machete*.snap --dangerous --classic
+  git machete version
+  sudo snap remove git-machete
+else
+  echo "MASTER"
+  echo "$SNAPCRAFT_LOGIN_CREDENTIALS_CONTENTS_BASE64" base64 -d > ~/.snapcraft.login
```
Comment (PawelLipski) [inline ci/snap/ci-deploy.sh]: Hmmm does it work? 😯 looks like it should be
```suggestion
  echo "$SNAPCRAFT_LOGIN_CREDENTIALS_CONTENTS_BASE64" | base64 -d > ~/.snapcraft.login
```

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -1,39 +1,37 @@
...
     source: .
     # We're setting the version via scriptlet rather than using `version: git`
     # since only pushing commits to master (and not pushing tags) triggers Snapcraft builds.
-    override-pull: |
+    override-build: |
       set -x
-      snapcraftctl pull
-      version=$(grep "__version__ = " git_machete/__init__.py | cut -d\' -f2)
+      snapcraftctl build
+      version="$(python3 setup.py --version)"
```
Comment (PawelLipski) [inline snap/snapcraft.yaml:29]: Good idea! lemme open a related issue...

File: snap/snapcraft.yaml [hunk 5]
```diff
@@ -1,39 +1,37 @@
 name: git-machete
-adopt-info: git-machete  # just for adopting `version`
 summary: git-machete is a versatile tool for organizing your git repo
 description: |
   git machete is a versatile tool for organizing your git repo, including features like:
   - Neat, customizable `git machete status` that shows what branches are in sync with their parent branch/remote tracking branch and which of them need to be rebased/merged/pulled/pushed
   - Semi-automatic `git machete traverse` through the branches that helps you effortlessly rebase/merge and push/pull the branches you care for
   - Automatic discovery of branch relations (`git machete discover`)
+adopt-info: git-machete # just for adopting `version`
```
Comment (PawelLipski) [inline snap/snapcraft.yaml]: Any specific reason for these reorderings?

Comment (amalota) [reply snap/snapcraft.yaml]: @PawelLipski No, i initially did not know that there is snapcraft.yaml file already in the repo so i created mine from the scratch and only later found out the old one, i can reorder it so it matches previous version.

File: ci/snap/ci-deploy.sh (hunk 3 above)
Comment (amalota) [reply ci/snap/ci-deploy.sh]: @PawelLipski Yes, i am fairly certain it works because if it didn't it would throw errors while trying to log in.

Comment (PawelLipski) [reply ci/snap/ci-deploy.sh]: 😯

Comment (PawelLipski) [reply ci/snap/ci-deploy.sh]: Even without the pipe `|`? in the current shape, `base64 -d` are just loose params of `echo`, appended verbatim to `.snapcraft.login` 🤔

Comment (PawelLipski) [reply ci/snap/ci-deploy.sh]: ...which means that `$SNAPCRAFT_LOGIN_CREDENTIALS_CONTENTS_BASE64` are NOT decoded from base64...

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -23,17 +23,15 @@ license: MIT
...
-      snapcraftctl pull
-      version=$(grep "__version__ = " git_machete/__init__.py | cut -d\' -f2)
+      snapcraftctl build
+      version="$(python3 setup.py --version)"
       snapcraftctl set-version "$version"
 
 apps:
   git-machete:
     command: bin/git-machete
-    adapter: full
```
Comment (PawelLipski) [inline snap/snapcraft.yaml:39]: Without that... can the installed snap access network or files in the home directory? 🤔

Comment (amalota) [reply snap/snapcraft.yaml:39]: `Adapter's` default value is `full`, there is no need for this anymore

Comment (PawelLipski) [reply snap/snapcraft.yaml:39]: Okay! didn't remember that

**Review by PawelLipski — COMMENTED** [review]
Okay! looks good, pls remove all the temporary stuff

File: ci/snap/README.md [hunk 7]
```diff
@@ -1,3 +1 @@
 Note that this directory is distinct from top-level [snap/](https://github.com/VirtusLab/git-machete/tree/master/snap) directory.
```
Comment (PawelLipski) [inline ci/snap/README.md:1]: In terms of READMEs... pls also update CONTRIBUTING.md accordingly!

**Comment by amalota** [comment]
@PawelLipski Can you install machete using snap from edge channel and test if it works properly, thanks.

**Comment by PawelLipski** [comment]
> @PawelLipski Can you install machete using snap from edge channel and test if it works properly, thanks.

yes

**Review by PawelLipski — APPROVED** [review]
Okay! let's see how it works with 3.8.0 🤞🏻

**Pull request merged**